Scale implementation

duncanmalashock opened the pull request:
References for intervals used:
https://en.wikipedia.org/wiki/Synthetic_mode
https://en.wikipedia.org/wiki/Hexatonic_scale#Whole_tone_scale
https://en.wikipedia.org/wiki/Pentatonic_scale#Types
https://en.wikipedia.org/wiki/Modes_of_limited_transposition

duncanmalashock committed:
ScaleClasses with at least one example of each type

duncanmalashock committed:
Scale and scale function

duncanmalashock commented:
Overall I think this is a good model of `ScaleClass`es, and of the `PitchClass`es that make up a `Scale`. But it is unclear how `Scale`s can be used effectively this way, since there is no concept of a "current degree" that can be used to move up and down the scale. I will consider options for this.

duncanmalashock commented:
Thinking about interface for getting notes from a scale:

One thing I like about the way `Scale`s are modeled is that we make very few assumptions. Pentatonic scales have five notes, hexatonic scales have six, etc. So there is no concept of trying to get the "sixth degree" from a scale that doesn't have one. I think this is good.

Having said that, we will need a way of stepping up and down the scale. Currently we don't have one. Should the `Scale` be responsible for this?

The least we could do, probably, is to expose a `toList` function and let client code handle the list traversal to find the notes they wanted. But I don't think this is a good option, because it requires that client code needs to understand that `Scale`s return to the first note when they reach the end, which is a domain concept that we should model.

Also, there are questions about `Scale`s of `PitchClass`es vs. `Scale`s of `Pitch`es (with an octave):
- Are they two separate types of `Scale`s, like `PitchScale` and `PitchClassScale`?
- A `Pitch` at the end of a `Scale` should increase its octave. How should that behavior be modeled?

battermann commented on src/MusicTheory/ScaleClass.elm:
```diff
@@ -0,0 +1,165 @@
+
+
+diminishedWholeHalf : ScaleClass
+diminishedWholeHalf =
```
Another option I had in mind is to have different models for diatonic and non-diatonic scales.

Non-diatonic scales could be defined as semitones from the root or as stacked semitones.

Not sure if this makes sense because intervals can always be converted to semitones (but not vice versa).

But in terms of domain this might be more accurate.

duncanmalashock replied:
I was able to find exact intervals from the Wikipedia sites in the PR description, so I think if we're able to continue finding intervals reliably, we should keep doing so.

If we end up having to define scales in terms of semitones instead, that could really introduce a lot of complication in how we represent scale steps, since, as you know, the pitch class N semitones away from the first pitch class has infinite possible spellings.

duncanmalashock replied:
But, modeling diatonic and non-diatonic scales separately may turn out to be a good idea! We've already made the distinction between diatonic scales having modes and non-diatonic scales not having them. And there are probably other distinctions we will identify.

I just personally think that it would be preferable to keep using intervals, rather than semitones.

battermann replied:
Let's keep intervals then.

battermann commented on src/MusicTheory/ScaleClass.elm:
```diff
@@ -0,0 +1,165 @@
+ionian : ScaleClass
+ionian =
+    HeptatonicScaleClass
+        { firstDegree = Interval.perfectUnison
```
Do we need an explicit first degree? Will the first degree ever be anything other than a perfect unison?

duncanmalashock replied:
I was curious about whether there are any cases of that, and I suspect there aren’t.

battermann commented on src/MusicTheory/ScaleClass.elm:
```diff
@@ -0,0 +1,165 @@
+
+
+minor : ScaleClass
+minor =
```
So far I don't have a solution to this, but in my ideal implementation, all the church modes should be derivable from the ionian scale.

This is easier with stacked intervals I think, but that has other disadvantages.

Could there be a type `Mode`, which is a `ScaleClass` with a starting degree?

duncanmalashock replied:
Can you be more specific about what you want modes for?

The simplest way to get modes (of the major and melodic minor, for example) would be to define them here in this file as `ScaleClass`es. I think maybe that's the way we should start.

As far as I know, there is no such thing as a mode of a non-diatonic scale. So if we wrote a `modeOf` function that took a scale degree, we would probably have to either:
1. model diatonic scales and non-diatonic scales separately
2. return a `Result` for handling errors in the case where the `Scale` is not diatonic

Is there a reason you want a `Mode` type? Would a `Mode` work the same way as a scale?

battermann replied:
> Can you be more specific about what you want modes for?

Not really. I just thought it would be nice to derive them, because that reflects the domain.

But I guess if we hard code all the scales, that's fine, too.

duncanmalashock replied:
I agree, it would be nice!

I think if we model diatonic scales separately (which I am leaning toward now), we could write:
```
modeOf : DiatonicScale -> DiatonicScaleDegree -> DiatonicScale
```
and
```
isModeOf : DiatonicScale -> DiatonicScale -> Bool
```
Which could be nice for deriving and detecting those relationships.

battermann commented on src/MusicTheory/ScaleClass.elm:
```diff
@@ -0,0 +1,165 @@
+    , HexatonicScaleClassIntervals
+    , OctatonicScaleClassIntervals
+    , PentatonicScaleClassIntervals
+    , ScaleClass(..)
```
We should make this opaque and define some constructors.

duncanmalashock replied:
Look at how these constructors are used in Scale.scale. If you don’t want to expose them, we could merge ScaleClass and Scale into one module.

duncanmalashock replied:
Also, I am all in favor of making types opaque and limiting the public API, but I think we don't need to be so strict about it just yet. In some cases, like this one, modules may need to expose some internals to each other. But we have a "last line of defense" in the `elm.json` file's list of `exposed-modules`, where we can define which modules get exposed to callers outside. If I remember correctly, sometimes package authors will define "wrapper" modules that only expose what they want exposed, and not include the others in `exposed-modules`. I can try to find some examples of that.

battermann replied:
I think `ScaleClass` should be exposed, at least at some point.

Clients might want to define their own scale classes. There should be dedicated constructors that validate the input and return `Result` or something similar.

But this could be for the future.

battermann reviewed (commented):
Looks nice! I made some comments, but I'm not done yet.

battermann reviewed (commented):
This all looks good.

The only thing I would like to be changed is that the constructors for scale classes are hidden.

ATM you could create scale classes that do not make sense.

battermann commented:
> Having said that, we will need a way of stepping up and down the scale. Currently we don't have one. Should the Scale be responsible for this?

Exposing `toList` would be good and I think enough for now. Maybe we should ignore the edge case of different ascending and descending definitions. If a client really needs it, they can build it themselves with the given model somehow.

battermann commented:
> Are they two separate types of Scales, like PitchScale and PitchClassScale?

PitchScale doesn't make sense to me. This is rather a `List Pitch`.

We could have something like:

`scaleWithin : Range -> Scale -> List Pitch`

duncanmalashock commented:
> > Take a look at https://github.com/elm-music/music-theory/pull/24/files#diff-5d144c41e871f3f6158c8e8b1310830dR62 and you'll see that the constructors for ScaleClass need to get used by Scale.scale, in order to determine what kind of ScaleClass it is and destructure its intervals.
> 
> Can we define them in a single module?

We can, by moving `ScaleClass` and `Scale` into the same module. But I think it's better to keep them separate, so that we can think about each domain idea, instead of putting them together to protect against unwanted access, for two reasons:
1. We haven't published the package yet, so there aren't yet any users who could use it!
2. When we're ready to publish, we can choose which files to expose. At that point, we can make more specific decisions about how to structure the files to expose the right API, by importing and exposing in new public-facing modules.

duncanmalashock commented:
Now that I think about it more, it doesn't seem like such a big problem to move the implementation of `ScaleClass` into the same file as `Scale`. I am okay with doing that.

duncanmalashock commented:
Going to make three revisions today based on our conversation:
1. Remove `firstDegree` from all scale interval types and store the root elsewhere on the `Scale`
2. Move `ScaleClass` into the `Scale` module to avoid exposing `ScaleClass`'s variants
3. Implement `Scale.toList`

duncanmalashock committed:
Revisions

- Remove firstDegree from all scale interval types and store the root elsewhere on the Scale
- Move ScaleClass into the Scale module to avoid exposing ScaleClass's variants
- Implement Scale.toList
